### PR TITLE
I029 - Code Style Checker

### PR DESCRIPTION
**Pull request by chenshuiluke**
You can run `python3 code_style.py --test` to run it against the test script I made for it:

	Found Def!
	Recognized arguments for function __init__: ['self', 'config', 'host']
	Recognized returns for function __init__: []
	Docstring arguments: ['config', 'host']
	Docstring returns: []

	============****************************************============

	Found Def!
	Recognized arguments for function ethernet: ['self']
	Recognized returns for function ethernet: ['html']
	Docstring arguments: ['config', 'host']
	The function signature argument list for ethernet and its docstring argument list aren't of the same length
	Docstring returns: ['html']

	============****************************************============

	Found Def!
	Recognized arguments for function device: ['self']
	Recognized returns for function device: ['html']
	Docstring arguments: ['config', 'host']
	The function signature argument list for device and its docstring argument list aren't of the same length
	Docstring returns: ['html']

	============****************************************============

	Found Def!
	Recognized arguments for function __init__: ['self', 'queue']
	Recognized returns for function __init__: []
	Docstring arguments: ['queue']
	Docstring returns: []

	============****************************************============

	Found Def!
	Recognized arguments for function run: ['self']
	Recognized returns for function run: []
	Docstring arguments: []
	Docstring returns: []

	============****************************************============

	Found Def!
	Recognized arguments for function _port_enabled: ['port_data']
	Recognized returns for function _port_enabled: ['enabled']
	Docstring arguments: ['port_data']
	Docstring returns: ['active']
	**********enabled is returned in _port_enabled, but is not found in the return section of the docstring**********

	============****************************************============

	Recognized arguments for function _port_up: ['port_data']
	Recognized returns for function _port_up: ['enabled']
	Docstring arguments: ['port_data']
	Docstring returns: ['active']
	**********enabled is returned in _port_up, but is not found in the return section of the docstring**********

	============****************************************============

	Recognized arguments for function _get_state: ['port_data']
	Recognized returns for function _get_state: ['state']
	Docstring arguments: ['port_data']
	Docstring returns: ['state']

	============****************************************============

	Recognized arguments for function _get_inactive: []
	Recognized returns for function _get_inactive: ["'TBD'"]
	Docstring arguments: ['port_data']
	The function signature argument list for _get_inactive and its docstring argument list aren't of the same length
	Docstring returns: ['inactive']
	**********'TBD' is returned in _get_inactive, but is not found in the return section of the docstring**********

	============****************************************============

	Recognized arguments for function _get_speed: ['port_data']
	Recognized returns for function _get_speed: ['speed']
	Docstring arguments: ['port_data']
	Docstring returns: ['speed']

	============****************************************============

	Recognized arguments for function _get_cdp: ['port_data']
	Recognized returns for function _get_cdp: ['value']
	Docstring arguments: ['port_data']
	Docstring returns: ['value']

	============****************************************============

	Recognized arguments for function _get_lldp: ['port_data']
	Recognized returns for function _get_lldp: ['value']
	Docstring arguments: ['port_data']
	Docstring returns: ['value']

	============****************************************============

	Recognized arguments for function _get_duplex: ['port_data']
	Recognized returns for function _get_duplex: ['duplex']
	Docstring arguments: ['port_data']
	Docstring returns: ['duplex']

	============****************************************============

	Recognized arguments for function _get_vlan: ['port_data']
	Recognized returns for function _get_vlan: ['vlans']
	Docstring arguments: ['port_data']
	Docstring returns: []
	**********vlans is returned in _get_vlan, but is not found in the return section of the docstring**********

	============****************************************============

	Recognized arguments for function _html_footer: []
	Recognized returns for function _html_footer: []
	Docstring arguments: []
	Docstring returns: ['html']

	============****************************************============

	Found Def!
	Recognized arguments for function _html_header: ['host']
	Recognized returns for function _html_header: []
	Docstring arguments: ['host']
	Docstring returns: ['html']

	============****************************************============

	Found Def!
	Recognized arguments for function _index_html: ['config']
	Recognized returns for function _index_html: []
	Docstring arguments: []
	The function signature argument list for _index_html and its docstring argument list aren't of the same length
	**********config is found in _index_html arglist, but not in docstring**********
	Docstring returns: ['html']

	============****************************************============

	Found Def!
	Recognized arguments for function _port_table: ['data_dict']
	Recognized returns for function _port_table: ['output']
	Docstring arguments: ['data_dict']
	Docstring returns: ['output']

	============****************************************============

	Recognized arguments for function _device_table: ['data_dict']
	Recognized returns for function _device_table: ['output']
	Docstring arguments: ['data_dict']
	Docstring returns: ['output']

	============****************************************============

	Recognized arguments for function _uptime: ['seconds']
	Recognized returns for function _uptime: ['result']
	Docstring arguments: ['seconds']
	Docstring returns: ['result']

	============****************************************============

	Recognized arguments for function cdpcachedeviceid: ['self', 'var1', 'oidonly']
	Recognized returns for function cdpcachedeviceid: ['oid', 'data_dict']
	Docstring arguments: ['var1', 'oidonly', 'var']
	The function signature argument list for cdpcachedeviceid and its docstring argument list aren't of the same length
	Docstring returns: ['data_dict']
	**********oid is returned in cdpcachedeviceid, but is not found in the return section of the docstring**********

	============****************************************============

The code needs some major refactoring. I suppose it's ironic that this code style checker doesn't really adhere to any code styles yet, but I will work on that.

### CODE REVIEW

File: code_style_test.py [hunk 1]
```diff
@@ -0,0 +1,661 @@
+#!usr/bin/env python3
```
Comment (Pr0x1m4) [inline code_style_test.py]: Why is this file being included in the pull request?

File: code_style.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
+#!/usr/bin/python
```
Comment (Pr0x1m4) [inline code_style.py]: Possible to rename to lint.py, also update README.md with instructions on how to use

Comment (palisadoes) [inline code_style.py]: In addition to the renaming, the code should only provide output if issues are found. For example missing arguments in doc string explanations, arguments in doc string explanations that are out of order when compared to how they are called in the function / method's "def".

Line numbers for where issues are found should be shown too. Emulate pylint output as much as possible.

**Comment by chenshuiluke** [comment]
@Pr0x1m4 @palisadoes Made the output more concise and only have error information. the file `code_style_test.py` is there only to make it easy to test the script. I will remove it once you guys deem the PR suitable.

output of `python lint.py --test`

    ethernet: no. of arguments in function signature and no. of arguments in docstring are not the same.
    device: no. of arguments in function signature and no. of arguments in docstring are not the same.
    _port_enabled: enabled is returned in the function, but is not found in the return section of the docstring
    _port_enabled: active is in the return section of the docstring, but is not returned in the function
    _port_up: enabled is returned in the function, but is not found in the return section of the docstring
    _port_up: active is in the return section of the docstring, but is not returned in the function
    _get_inactive: no. of arguments in function signature and no. of arguments in docstring are not the same.
    _get_inactive: 'TBD' is returned in the function, but is not found in the return section of the docstring
    _get_inactive: inactive is in the return section of the docstring, but is not returned in the function
    _get_vlan: vlans is returned in the function, but is not found in the return section of the docstring
    _html_footer: html is in the return section of the docstring, but is not returned in the function
    _html_header: html is in the return section of the docstring, but is not returned in the function
    _index_html: no. of arguments in function signature and no. of arguments in docstring are not the same.
    _index_html: config is found in function signature's argslist, but not in docstring
    _index_html: html is in the return section of the docstring, but is not returned in the function
    cdpcachedeviceid: no. of arguments in function signature and no. of arguments in docstring are not the same.
    cdpcachedeviceid: oid is returned in the function, but is not found in the return section of the docstring

**Comment by chenshuiluke** [comment]
Added instructions to README.md

**Comment by palisadoes** [comment]
Luke, please add the line number in the processed file of where each fault was found to the output. This will improve the user's speed in correcting it.

**Comment by chenshuiluke** [comment]
Okay, line numbers of the functions have been added:

    __init__: line 27: no. of arguments in function signature and no. of arguments in docstring are not the same.
    __init__: line 27: config is found in function signature's argslist, but not in docstring
    __init__: line 27: host is found in function signature's argslist, but not in docstring
    __init__: line 27: no docstring found
    ethernet: line 33: no. of arguments in function signature and no. of arguments in docstring are not the same.
    ethernet: line 33: config is found in docstring, but not in function signature's argslist
    ethernet: line 33: host is found in docstring, but not in function signature's argslist
    device: line 50: no. of arguments in function signature and no. of arguments in docstring are not the same.
    device: line 50: config is found in docstring, but not in function signature's argslist
    device: line 50: host is found in docstring, but not in function signature's argslist
    _port_enabled: line 221: enabled is returned in the function, but is not found in the return section of the docstring
    _port_enabled: line 221: active is in the return section of the docstring, but is not returned in the function
    _port_up: line 244: enabled is returned in the function, but is not found in the return section of the docstring
    _port_up: line 244: active is in the return section of the docstring, but is not returned in the function
    _get_inactive: line 290: no. of arguments in function signature and no. of arguments in docstring are not the same.
    _get_inactive: line 290: port_data is found in docstring, but not in function signature's argslist
    _get_inactive: line 290: 'TBD' is returned in the function, but is not found in the return section of the docstring
    _get_inactive: line 290: inactive is in the return section of the docstring, but is not returned in the function
    _get_vlan: line 412: vlans is returned in the function, but is not found in the return section of the docstring
    _html_footer: line 440: html is in the return section of the docstring, but is not returned in the function
    _html_header: line 458: html is in the return section of the docstring, but is not returned in the function
    _index_html: line 482: no. of arguments in function signature and no. of arguments in docstring are not the same.
    _index_html: line 482: config is found in function signature's argslist, but not in docstring
    _index_html: line 482: html is in the return section of the docstring, but is not returned in the function
    cdpcachedeviceid: line 624: no. of arguments in function signature and no. of arguments in docstring are not the same.
    cdpcachedeviceid: line 624: var is found in docstring, but not in function signature's argslist
    cdpcachedeviceid: line 624: oid is returned in the function, but is not found in the return section of the docstring

**Comment by chenshuiluke** [comment]
I opted to just print the line of the function for the errors because they almost always involve the argslist of the function signature or the docstring, which will generally be found on the same line as the function signature or just a few lines after it. I could try and find a way to print more specific line numbers for the errors if you want.

**Comment by chenshuiluke** [comment]
@Pr0x1m4 I added pep8 and pyflakes output as you requested. Output from `python lint.py --test`:

    pep8 output:
    code_style_test.py:625:1: E302 expected 2 blank lines, found 1


    pyflakes output:
    code_style_test.py:635: undefined name 'defaultdict'
    code_style_test.py:647: undefined name '_ifindex'


    Script output:
    __init__: line 28: no. of arguments in function signature and no. of arguments in docstring are not the same.
    __init__: line 28: config is found in function signature's argslist, but not in docstring
    __init__: line 28: host is found in function signature's argslist, but not in docstring
    __init__: line 28: no docstring found
    ethernet: line 34: no. of arguments in function signature and no. of arguments in docstring are not the same.
    ethernet: line 34: config is found in docstring, but not in function signature's argslist
    ethernet: line 34: host is found in docstring, but not in function signature's argslist
    device: line 51: no. of arguments in function signature and no. of arguments in docstring are not the same.
    device: line 51: config is found in docstring, but not in function signature's argslist
    device: line 51: host is found in docstring, but not in function signature's argslist
    _port_enabled: line 222: enabled is returned in the function, but is not found in the return section of the docstring
    _port_enabled: line 222: active is in the return section of the docstring, but is not returned in the function
    _port_up: line 245: enabled is returned in the function, but is not found in the return section of the docstring
    _port_up: line 245: active is in the return section of the docstring, but is not returned in the function
    _get_inactive: line 291: no. of arguments in function signature and no. of arguments in docstring are not the same.
    _get_inactive: line 291: port_data is found in docstring, but not in function signature's argslist
    _get_inactive: line 291: 'TBD' is returned in the function, but is not found in the return section of the docstring
    _get_inactive: line 291: inactive is in the return section of the docstring, but is not returned in the function
    _get_vlan: line 413: vlans is returned in the function, but is not found in the return section of the docstring
    _html_footer: line 441: html is in the return section of the docstring, but is not returned in the function
    _html_header: line 459: html is in the return section of the docstring, but is not returned in the function
    _index_html: line 483: no. of arguments in function signature and no. of arguments in docstring are not the same.
    _index_html: line 483: config is found in function signature's argslist, but not in docstring
    _index_html: line 483: html is in the return section of the docstring, but is not returned in the function
    cdpcachedeviceid: line 625: no. of arguments in function signature and no. of arguments in docstring are not the same.
    cdpcachedeviceid: line 625: var is found in docstring, but not in function signature's argslist
    cdpcachedeviceid: line 625: oid is returned in the function, but is not found in the return section of the docstring

**Comment by chenshuiluke** [comment]
okay added filename to output:

    pep8 output:
    code_style_test.py:625:1: E302 expected 2 blank lines, found 1


    pyflakes output:
    code_style_test.py:635: undefined name 'defaultdict'
    code_style_test.py:647: undefined name '_ifindex'


    Script output:
    code_style_test.py: __init__: line 28: no. of arguments in function signature and no. of arguments in docstring are not the same.
    code_style_test.py: __init__: line 28: config is found in function signature's argslist, but not in docstring
    code_style_test.py: __init__: line 28: host is found in function signature's argslist, but not in docstring
    code_style_test.py: __init__: line 28: no docstring found
    code_style_test.py: ethernet: line 34: no. of arguments in function signature and no. of arguments in docstring are not the same.
    code_style_test.py: ethernet: line 34: config is found in docstring, but not in function signature's argslist
    code_style_test.py: ethernet: line 34: host is found in docstring, but not in function signature's argslist
    code_style_test.py: device: line 51: no. of arguments in function signature and no. of arguments in docstring are not the same.
    code_style_test.py: device: line 51: config is found in docstring, but not in function signature's argslist
    code_style_test.py: device: line 51: host is found in docstring, but not in function signature's argslist
    code_style_test.py: _port_enabled: line 222: enabled is returned in the function, but is not found in the return section of the docstring
    code_style_test.py: _port_enabled: line 222: active is in the return section of the docstring, but is not returned in the function
    code_style_test.py: _port_up: line 245: enabled is returned in the function, but is not found in the return section of the docstring
    code_style_test.py: _port_up: line 245: active is in the return section of the docstring, but is not returned in the function
    code_style_test.py: _get_inactive: line 291: no. of arguments in function signature and no. of arguments in docstring are not the same.
    code_style_test.py: _get_inactive: line 291: port_data is found in docstring, but not in function signature's argslist
    code_style_test.py: _get_inactive: line 291: 'TBD' is returned in the function, but is not found in the return section of the docstring
    code_style_test.py: _get_inactive: line 291: inactive is in the return section of the docstring, but is not returned in the function
    code_style_test.py: _get_vlan: line 413: vlans is returned in the function, but is not found in the return section of the docstring
    code_style_test.py: _html_footer: line 441: html is in the return section of the docstring, but is not returned in the function
    code_style_test.py: _html_header: line 459: html is in the return section of the docstring, but is not returned in the function
    code_style_test.py: _index_html: line 483: no. of arguments in function signature and no. of arguments in docstring are not the same.
    code_style_test.py: _index_html: line 483: config is found in function signature's argslist, but not in docstring
    code_style_test.py: _index_html: line 483: html is in the return section of the docstring, but is not returned in the function
    code_style_test.py: cdpcachedeviceid: line 625: no. of arguments in function signature and no. of arguments in docstring are not the same.
    code_style_test.py: cdpcachedeviceid: line 625: var is found in docstring, but not in function signature's argslist
    code_style_test.py: cdpcachedeviceid: line 625: oid is returned in the function, but is not found in the return section of the docstring

**Comment by chenshuiluke** [comment]
@Pr0x1m4 I grouped the script output section by filename like you requested `python lint.py --test infoset/web/ws_device.py`:

    pep8 output:

    code_style_test.py:625:1: E302 expected 2 blank lines, found 1


    pyflakes output:

    code_style_test.py:635: undefined name 'defaultdict'
    code_style_test.py:647: undefined name '_ifindex'


    Script output:

    code_style_test.py: 
    	__init__: line 28: no. of arguments in function signature and no. of arguments in docstring are not the same.
    	__init__: line 28: config is found in function signature's argslist, but not in docstring
    	__init__: line 28: host is found in function signature's argslist, but not in docstring
    	__init__: line 28: no docstring found
    	ethernet: line 34: no. of arguments in function signature and no. of arguments in docstring are not the same.
    	ethernet: line 34: config is found in docstring, but not in function signature's argslist
    	ethernet: line 34: host is found in docstring, but not in function signature's argslist
    	device: line 51: no. of arguments in function signature and no. of arguments in docstring are not the same.
    	device: line 51: config is found in docstring, but not in function signature's argslist
    	device: line 51: host is found in docstring, but not in function signature's argslist
    	_port_enabled: line 222: enabled is returned in the function, but is not found in the return section of the docstring
    	_port_enabled: line 222: active is in the return section of the docstring, but is not returned in the function
    	_port_up: line 245: enabled is returned in the function, but is not found in the return section of the docstring
    	_port_up: line 245: active is in the return section of the docstring, but is not returned in the function
    	_get_inactive: line 291: no. of arguments in function signature and no. of arguments in docstring are not the same.
    	_get_inactive: line 291: port_data is found in docstring, but not in function signature's argslist
    	_get_inactive: line 291: 'TBD' is returned in the function, but is not found in the return section of the docstring
    	_get_inactive: line 291: inactive is in the return section of the docstring, but is not returned in the function
    	_get_vlan: line 413: vlans is returned in the function, but is not found in the return section of the docstring
    	_html_footer: line 441: html is in the return section of the docstring, but is not returned in the function
    	_html_header: line 459: html is in the return section of the docstring, but is not returned in the function
    	_index_html: line 483: no. of arguments in function signature and no. of arguments in docstring are not the same.
    	_index_html: line 483: config is found in function signature's argslist, but not in docstring
    	_index_html: line 483: html is in the return section of the docstring, but is not returned in the function
    	cdpcachedeviceid: line 625: no. of arguments in function signature and no. of arguments in docstring are not the same.
    	cdpcachedeviceid: line 625: var is found in docstring, but not in function signature's argslist
    	cdpcachedeviceid: line 625: oid is returned in the function, but is not found in the return section of the docstring

    infoset/web/ws_device.py: 
    	ethernet: line 44: no. of arguments in function signature and no. of arguments in docstring are not the same.
    	ethernet: line 44: config is found in docstring, but not in function signature's argslist
    	ethernet: line 44: host is found in docstring, but not in function signature's argslist
    	device: line 61: no. of arguments in function signature and no. of arguments in docstring are not the same.
    	device: line 61: config is found in docstring, but not in function signature's argslist
    	device: line 61: host is found in docstring, but not in function signature's argslist
    	_port_enabled: line 232: enabled is returned in the function, but is not found in the return section of the docstring
    	_port_enabled: line 232: active is in the return section of the docstring, but is not returned in the function
    	_port_up: line 255: enabled is returned in the function, but is not found in the return section of the docstring
    	_port_up: line 255: active is in the return section of the docstring, but is not returned in the function
    	_get_inactive: line 301: no. of arguments in function signature and no. of arguments in docstring are not the same.
    	_get_inactive: line 301: port_data is found in docstring, but not in function signature's argslist
    	_get_inactive: line 301: 'TBD' is returned in the function, but is not found in the return section of the docstring
    	_get_inactive: line 301: inactive is in the return section of the docstring, but is not returned in the function
    	_get_vlan: line 423: vlans is returned in the function, but is not found in the return section of the docstring
    	_html_footer: line 451: html is in the return section of the docstring, but is not returned in the function
    	_html_header: line 469: html is in the return section of the docstring, but is not returned in the function
    	_index_html: line 493: no. of arguments in function signature and no. of arguments in docstring are not the same.
    	_index_html: line 493: config is found in function signature's argslist, but not in docstring
    	_index_html: line 493: html is in the return section of the docstring, but is not returned in the function

**Comment by Pr0x1m4** [comment]
Code looks great, merging to master and integrating into TravisCI

**Comment by palisadoes** [comment]
Luke, 

I looked at the code in more detail. There are a number of issues that need to be resolved:

1. The code doesn't pass pylint, pep8 and pep257. 
2. There are no comments to give an idea of what it does.
3. There is no help (if required)
4. In other words, the code won't pass itself.

Please review and make the corrections.

**Pull request merged**